<!DOCTYPE html>
<html lang="en-US" xmlns="http://www.w3.org/1999/xhtml">
<head>
  <title>PhyloCorrelate Homepage</title>
  <style type="text/css">
    body, html {
      font-family: Helvetica, Arial, sans-serif;
      background-color: white;
      color: #114;
      margin: 0;
      padding: 0;
    }
    a {
      text-decoration: none;
    }
    a:hover {
      text-decoration: underline;
    }
    #titleBar {
      height: 80px;
      background-color: #cccccc;
      overflow: hidden;
      border-bottom: 5px solid black;
      border-left: 5px solid black;
      border-right: 5px solid black;
      border-top: 5px solid black;
    }
    #titleBar #container {
      margin-top: 14px;
    }
    #titleBar h1 {
      margin: 0 auto .5em auto;
      padding: .2em;
      color: black;
      text-align: center;
    }
    #outer-content {
      max-width: 910px;
      margin-left: auto;
      margin-right: auto;
    }
    #content {
      margin: 1em auto 1em auto;
      float: left;
    }
    #main{
      float: left;
      line-height: 24px;
    }
    #inner {
      width:50%;
      margin: 0 auto;
      margin-right: 190px;
    }
    .button {
      background-color: #4ca64c;
      border: 3px solid black;
      color: white;
      padding: 15px 25px;
      text-align: center;
      font-size: 16px;
      cursor: pointer;
      border-radius: 30px;
    }
    .button:hover {
      background-color: grey;
    }
    .boxed {
      border: 1px solid #d3d3d3;
      background-color: #FFF;
      width: auto;
      margin: 0 auto;
      padding: 20px;
    }
    .footer {
      position: fixed;
      left: 0;
      right: 0;
      bottom: 0;
      width: 100%;
      background-color: #FFF;
      color: grey;
      padding: 20px;
      width: 97%;
    }
  </style>
</head>
<body>
  <div id="titleBar">
    <div id="container">
      <h1>PhyloCorrelate</h1>
    </div>
  </div>
  <div id="outer-content">
    <div id="content">
      <div id="main">

        <h1>Overview</h1>

        <p>Phylocorrelate is a tool that detects pairs of gene/protein families with similar phylogenetic distributions across the bacterial tree of life. The database contains pre-computed phylogenomic correlations for 27,372 gene families across 28,315 species. PhyloCorrelate uses the GTDB bacterial tree and gene occurrences from AnnoTree including entries from KEGG, PFAM, and TIGRFAM.</p>

        <p>Gene co-occurrence is important to detect as it may suggest an interaction between the genes, membership in the same pathway or complex, or more general functional associations or dependencies.</p>

        <br>
        <div id="wrapper" style = "text-align: center;">
          <p><img src="./logo.png" height="126" width="180"></p>
        </div>

        <div id="wrapper" style = "text-align: center;">
          <a href="./app/"><button class="button" style="width:180px;"><b>Launch</b></button></a>
        </div>

        <div id="inner">
          <p style="width:400px;text-align:justify;">Explore phylogenomic correlations for existing genes/proteins of interest, or submit new genes for co-occurrence analysis.</p>
        </div>

          <br></br>

          <h1>How does it work?</h1>

          <p>PhyloCorrelate detects pairs of gene/protein families with similar phylogenetic distributions across the bacterial tree of life. To find genes with similar distributions to a query gene (e.g., cobD below), first we examine the presence/absence distribution of the query gene across the tree. We then examine the distributions of all other genes and calculate their similarity using several metrics, including the Jaccard overlap coefficient (JC) and hypergeometric p-value statistic. The genes with the highest JC scores will be the top candidate correlation partners.</p>

          <p>But it isnâ€™t quite so simple. PhyloCorrelate also implements a phylogenetic correction to ensure that co-occurrence patterns are statistically significant given the topology of the tree. This helps to remove apparent correlations that occur only because gene A and B happen to be conserved within the same lineages.</p>

          <div id="wrapper" style = "text-align: center;">
            <p><img src="./phylocorrelateSummary.jpg" style="max-width:900px;width:100%;"></p>
          </div>

          <h1>Team</h1>

          <ul>
            <li>Benjamin J.-M. Tremblay: Lead developer and bioinformatics technician</li>
            <li>Briallen Lobb: Case studies and tool design</li>
            <li>Andrew C. Doxey: PI and team supervision</li>
          </ul>

          <h1>Cite us</h1>

        <p>If you use PhyloCorrelate in a manuscript, please cite the following:</p>
        <ol type="1">
          <li>Tremblay et al. (2021) PhyloCorrelate: inferring bacterial gene-gene functional associations through large-scale phylogenetic profiling. Bioinformatics, <a href="https://doi.org/10.1093/bioinformatics/btaa1105">https://doi.org/10.1093/bioinformatics/btaa1105</a>.</li>
          <li>Mendler et al. (2019) AnnoTree: visualization and exploration of a functionally annotated microbial tree of life. <i>Nucleic Acids Research</i>, Volume 47, Issue 9, 21 May 2019, Pages 4442-4448, <a href="https://doi.org/10.1093/nar/gkz246">https://doi.org/10.1093/nar/gkz246</a></li>
        </ol>

        <h1>Source code/data</h1>

        <p>The source code for the PhyloCorrelate R shiny app is available at <a href="https://github.com/doxeylab/phylocorrelate">https://github.com/doxeylab/phylocorrelate</a>. A standalone R package implementation is also available at <a href="https://github.com/bjmt/phylocorrpkg">https://github.com/bjmt/phylocorrpkg</a>.</p>

      </div>
    </div>
  </div>
</body>
</html>
